aws-datastore/src/main/java/com/amplifyframework/datastore/syncengine/OutboxMutationFailedEvent.java

```diff
@@ -0,0 +1,183 @@
+    @NonNull
+    public MutationErrorType getErrorType() {
+        return errorType;
+    }
+
+    /**
+     * Gets the mutation type.
+     * @return the mutation type
+     */
+    @NonNull
+    public MutationType getOperation() {
+        return operation;
+    }
+
+    /**
+     * Gets the local model that failed to be mutated.
+     * @return the local model
+     */
+    @NonNull
+    public M getModel() {
+        return model;
+    }
+
+    /**
+     * Gets the name of the model.
+     * @return the model name
+     */
+    @NonNull
+    public String getModelName() {
+        return modelName;
+    }
+
+    @Override
+    public HubEvent<OutboxMutationFailedEvent<M>> toHubEvent() {
+        return HubEvent.create(DataStoreChannelEventName.OUTBOX_MUTATION_FAILED, this);
+    }
+
+    @Override
+    public String toString() {
+        return "OutboxMutationFailedEvent{" +
+                "errorType=" + errorType +
+                ", operation=" + operation +
+                ", modelName=" + modelName +
+                ", model=" + model +
+                '}';
+    }
+
+    /**
+     * Categorization of error types that caused a mutation publication to fail.
+     */
+    public enum MutationErrorType {
+        /**
+         * The mutation operation is not authorized for the user.
+         */
+        UNAUTHORIZED("Unauthorized"),
+
+        /**
+         * Fallback type for any error that is yet to be categorized.
+         */
+        UNKNOWN("Unknown");
+
+        private final String value;
+
+        MutationErrorType(String value) {
+            this.value = value;
+        }
+
+        /**
+         * Get the value of error type received from cloud.
+         * @return value of error type received from cloud
+         */
+        @NonNull
+        public String getValue() {
+            return value;
+        }
+
+        /**
+         * Returns an enum value of matching error type value.
+         * @param value the type of error that caused mutation publication failure
+         * @return An enum value of matching error type. If there is no match,
+         *          then return {@link MutationErrorType#UNKNOWN}.
+         */
+        @NonNull
+        public static MutationErrorType enumerate(@Nullable String value) {
+            for (MutationErrorType type : MutationErrorType.values()) {
+                if (type.getValue().equals(value)) {
+                    return type;
+                }
+            }
+            return UNKNOWN;
+        }
+
+        // Look at the first error to extract error type.
+        private static MutationErrorType fromGraphQlErrors(List<GraphQLResponse.Error> errors) {
+            try {
+                GraphQLResponse.Error firstError = errors.get(0);
+                AppSyncExtensions extensions = new AppSyncExtensions(firstError.getExtensions());
+                return enumerate(extensions.getErrorType().getValue());
+            } catch (RuntimeException parsingError) {
```

Comment:
It's generally best practice to catch the lowest level of exception that you need, not all `RuntimeException`.   Otherwise, it's unclear to future developers as to what _could_ go wrong.    Can you just do a null check (`Empty.check(firstError.getExtensions())`), and return `UNKNOWN` for that case?

Reply:
So this code snippet can throw error in following scenarios:
* `errors` is `null`
* `errors` is `empty`
* first item in `errors` is `null`
* error type inside first error object is `null`

Rather than excessively performing a null check, I decided to do a catch-all. However, I did fix the `RuntimeException` to something more specific (`NullPointerException | IndexOutOfBoundsException`) for clarification.